### Задание 1. Базовое изучение

Изучить набор данных и выбрать 8 столбцов (как числовые, так и категориальные). 

Провести расчет базовых метрик для них, кратко описать результаты.

In [1]:
import pandas as pd
import numpy as np

In [273]:
names = list(range(1, 29))
df = pd.read_csv('https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.csv', names=names, skiprows=1)
df.head()

1   2        3      4    5   6  7  8  9  10  ...     19    20 21    22 23  \
0  1   1   534817   39.2   88  20  ?  ?  4  1  ...     50    85  2     2  3   
1  2   1   530334  38.30   40  24  1  1  3  1  ...  33.00  6.70  ?     ?  1   
2  1   9  5290409  39.10  164  84  4  1  6  2  ...  48.00  7.20  3  5.30  2   
3  2   1   530255  37.30  104  35  ?  ?  6  2  ...  74.00  7.40  ?     ?  2   
4  2   1   528355      ?    ?   ?  2  1  3  1  ...      ?     ?  ?     ?  1   

  24    25 26 27 28  
0  2  2208  0  0  2  
1  2     0  0  0  1  
2  1  2208  0  0  1  
3  2  4300  0  0  2  
4  2     0  0  0  2  

[5 rows x 28 columns]

In [335]:
new_df = df[[2, 4, 5, 6, 7, 9, 10, 11, 13, 23]]
new_df.rename(columns={2: 'age', 4: 'rectal_temper', 5: 'pulse',  6: 'respir_rate', 7: 'temper_limbs', 9: 'mucous_membr', 10: 'time_fill_capill', 11: 'pain', 13: 'abdom_disten', 23: 'result'}, inplace=True)



/opt/anaconda3/lib/python3.7/site-packages/pandas/core/frame.py:4133: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


In [337]:
colums = ['age', 'rectal_temper', 'pulse', 'respir_rate', 'temper_limbs', 'mucous_membr', 'time_fill_capill', 'pain', 'abdom_disten', 'result']
dict_param = {} 

for i in colums:

    new_df[i] = pd.to_numeric(new_df[i], errors='coerce')
    
    count = new_df[i].count()
    mean = new_df[i].mean()
    mode = new_df[i].mode()[0]
    median = new_df[i].median()
    std = new_df[i].std()
    disp = new_df[i].var()
    minim = new_df[i].min()
    q25 = new_df[i].quantile(0.25)
    q75 = new_df[i].quantile(0.75)
    maxim = new_df[i].max()
    
    dict_param[i] = [count, mean, mode, std, disp, minim, q25, median, q75, maxim]

df_param = pd.DataFrame(dict_param)
df_param.index = ['count', 'mean', 'mode', 'std', 'disp', 'min', 'q25%', 'median', 'q75%', 'max']
df_param.head(15)


/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


age  rectal_temper       pulse  respir_rate  temper_limbs  \
count   299.000000     239.000000  275.000000   241.000000    243.000000   
mean      1.642140      38.166527   71.934545    30.427386      2.345679   
mode      1.000000      38.000000   48.000000    20.000000      3.000000   
std       2.177300       0.733508   28.680522    17.678256      1.046369   
disp      4.740634       0.538035  822.572342   312.520747      1.094888   
min       1.000000      35.400000   30.000000     8.000000      1.000000   
q25%      1.000000      37.800000   48.000000    18.000000      1.000000   
median    1.000000      38.200000   64.000000    24.000000      3.000000   
q75%      1.000000      38.500000   88.000000    36.000000      3.000000   
max       9.000000      40.800000  184.000000    96.000000      4.000000   

        mucous_membr  time_fill_capill        pain  abdom_disten      result  
count     253.000000        267.000000  244.000000    243.000000  298.000000  
mean        2.853755          1.303371    2.942623      2.259259    1.550336  
mode        1.000000          1.000000    3.000000      1.000000    1.000000  
std         1.620294          0.476623    1.303993      1.061472    0.737967  
disp        2.625353          0.227169    1.700398      1.126722    0.544596  
min         1.000000          1.000000    1.000000      1.000000    1.000000  
q25%        1.000000          1.000000    2.000000      1.000000    1.000000  
median      3.000000          1.000000    3.000000      2.000000    1.000000  
q75%        4.000000          2.000000    4.000000      3.000000    2.000000  
max         6.000000          3.000000    5.000000      4.000000    3.000000

*age* (Возраст) - полные, категориальные, 1 или 9, 9 редко: не вошло в первый квартиль 

*rectal_temper* (Ректальная температура) - 60 пропусков, численные, 35.4 до 40.8, разброс мал: среднее, мода и медиана - близки к значению 38

*pulse* (Пульс) - 24 пропуска, численные, 30 до 184, большой разброс, наиболее частое 48, по группе медиана выше нормы 64 (30-40 норма)

*respir_rate* (Частота дыхания) - 58 пропусков, численные, 8 до 96, значения сосредоточены в пределе 20-30, медиана 24 (выше нормы, норма 8-10), большие колебания: выбросы по максимальным значениям

*temper_limbs* (Температура конечностей) - 56 пропусков, категориальные, 1 до 4, наиболее частое 3 (показатель: круто), разброс мал: мода и медиана совпадают, часть лошадей имеют нормальную температуру (значения 1, 2)

*mucous_membr* (Слизистые оболочки) - 46 пропусков, категориальные, 1 до 6, большинство в норме (значения 1, 2): мода 1, лишь 25% имеют превышения нормы: третий квартиль - 3

*time_fill_capill* (Время наполнения капилляров) - 32 пропуска, категориальные, 1 до 3: что странно, в описании 1 и 2, 3 скорее всего выброс, показатель мало полезен: большинство лошадей имеют норму 

*pain* (Боль) - 55 пропусков, категориальные, 1 до 5, небольшой разброс: среднее, медиана и мода близки к значению 3, остальные значения равномерно распределены

*abdom_disten* (Вздутие живота) - 56 пропусков, категориальные, 1 до 4, наиболее частое значение 1 (норма), в целом данные распределены равномерно со смещением к меньшим значениям

*result* (Результат) - 1 пропуск, категориальные, 1 до 3, половина лошадей закончили плохо: средний 1.55, при медиане и моде 1

### Задание 2. Работа с выбросами

1. Найти выбросы 

2. Выдвинуть гипотезы об их причинах и проинтерпретировать результаты

3. Принять и обосновать решение о дальнейшей работе с ними

In [338]:
def quantiles(column):
    
    maxim = new_df[column].max()
    minim = new_df[column].min()
    q1 = new_df[column].quantile(0.25)
    q3 = new_df[column].quantile(0.75)
    iqr = q3 - q1
    if iqr == 0:
        iqr = 0.1
    lower_bound = q1 - (1.5 * iqr) 
    upper_bound = q3 + (1.5 * iqr)
    
    return q1, lower_bound, minim, q3, upper_bound, maxim, iqr

In [339]:
def presence_outliers(column):
    
    if (quantiles(column)[5] > quantiles(column)[4]) or (quantiles(column)[2] < quantiles(column)[1]):
        
        return True
    else:
        return False
    
    

In [341]:
def list_outliers(column):
    
    list_outliers_low = new_df[new_df[column].between(0, quantiles(column)[1], inclusive=True)]
    list_outliers_up = new_df[new_df[column].between(quantiles(column)[4], quantiles(column)[5], inclusive=True)]

    list_outliers = pd.concat([list_outliers_low, list_outliers_up])
    
    return list_outliers.sort_values(column)
    

In [342]:
colums_outliers = []

for column in colums:
    if presence_outliers(column):
        colums_outliers.append(column)
colums_outliers

['age', 'rectal_temper', 'pulse', 'respir_rate']

In [282]:
list_outliers(colums_outliers[0])

age  rectal_temper  pulse  respir_rate  temper_limbs  mucous_membr  \
2      9           39.1  164.0         84.0           4.0           6.0   
261    9           38.6   88.0         28.0           NaN           NaN   
254    9           38.8  184.0         84.0           1.0           1.0   
251    9           38.8  124.0         36.0           3.0           2.0   
243    9           38.2  124.0         88.0           1.0           2.0   
228    9           38.5  120.0         70.0           NaN           NaN   
226    9            NaN    NaN          NaN           NaN           NaN   
211    9           38.9  120.0         30.0           1.0           2.0   
190    9            NaN  100.0         44.0           2.0           1.0   
134    9           38.1  136.0         48.0           3.0           3.0   
108    9           38.4   84.0         40.0           3.0           2.0   
102    9           38.0  140.0         68.0           1.0           1.0   
81     9           38.1  100.0         80.0           3.0           2.0   
74     9           39.7  100.0          NaN           3.0           5.0   
73     9            NaN    NaN          NaN           NaN           NaN   
54     9           38.6  160.0         20.0           3.0           5.0   
40     9           39.0  150.0         72.0           NaN           NaN   
38     9           39.2  146.0         96.0           NaN           NaN   
22     9           38.3  130.0         60.0           NaN           NaN   
15     9            NaN  128.0         36.0           3.0           4.0   
12     9           38.0   92.0         28.0           1.0           2.0   
8      9           38.3   90.0          NaN           1.0           1.0   
274    9           38.8  150.0         50.0           1.0           6.0   
286    9           39.5   84.0         30.0           NaN           NaN   

     time_fill_capill  pain  abdom_disten  result  
2                 2.0   2.0           4.0     2.0  
261               NaN   NaN           NaN     1.0  
254               1.0   4.0           3.0     2.0  
251               1.0   2.0           4.0     2.0  
243               1.0   2.0           4.0     1.0  
228               NaN   NaN           NaN     1.0  
226               NaN   NaN           NaN     1.0  
211               2.0   3.0           3.0     1.0  
190               1.0   4.0           1.0     1.0  
134               1.0   5.0           3.0     2.0  
108               1.0   3.0           3.0     2.0  
102               1.0   3.0           2.0     1.0  
81                1.0   3.0           1.0     1.0  
74                2.0   2.0           NaN     3.0  
73                NaN   NaN           NaN     2.0  
54                1.0   3.0           4.0     2.0  
40                NaN   NaN           NaN     1.0  
38                NaN   NaN           NaN     2.0  
22                1.0   2.0           NaN     1.0  
15                2.0   4.0           3.0     2.0  
12                1.0   1.0           2.0     2.0  
8                 1.0   5.0           1.0     1.0  
274               2.0   5.0           2.0     2.0  
286               1.0   NaN           NaN     1.0

#### Комментарий по выбросам данных Возраст (age)

24 выброса по данным о молодых особях, из них 8 с частичными данными. 

Если бы данных было соразмерно со взрослыми особями, можно было бы оценить есть ли корреляция между данными в других столбцах и возрастом лошади. 

#### В данном случаем, считаю целесообразным: 

Игнорирование этого параметра.


In [283]:
list_outliers(colums_outliers[1])

age  rectal_temper  pulse  respir_rate  temper_limbs  mucous_membr  \
43     1           35.4  140.0         24.0           3.0           4.0   
140    1           36.0   42.0         30.0           NaN           5.0   
237    1           36.1   88.0          NaN           3.0           3.0   
79     1           36.4   98.0         35.0           3.0           4.0   
117    1           36.5   78.0         30.0           1.0           1.0   
297    1           36.5  100.0         24.0           3.0           3.0   
250    1           36.6   42.0         18.0           3.0           2.0   
98     1           39.6  108.0         51.0           3.0           6.0   
74     9           39.7  100.0          NaN           3.0           5.0   
19     1           39.9   72.0         60.0           1.0           5.0   
280    1           40.0   78.0          NaN           3.0           5.0   
53     1           40.3  114.0         36.0           3.0           1.0   
90     1           40.3  114.0         36.0           3.0           1.0   
258    1           40.8   72.0         42.0           3.0           1.0   

     time_fill_capill  pain  abdom_disten  result  
43                2.0   4.0           NaN     3.0  
140               1.0   NaN           NaN     2.0  
237               1.0   3.0           2.0     3.0  
79                1.0   4.0           2.0     2.0  
117               1.0   5.0           1.0     1.0  
297               1.0   3.0           3.0     1.0  
250               1.0   1.0           1.0     2.0  
98                2.0   2.0           3.0     1.0  
74                2.0   2.0           NaN     3.0  
19                2.0   5.0           4.0     1.0  
280               1.0   2.0           1.0     2.0  
53                2.0   2.0           3.0     3.0  
90                2.0   2.0           3.0     2.0  
258               1.0   2.0           1.0     2.0

#### Комментарий по выбросам данных Ректальная температура (rectal_temper)

14 выбросов по данным, половина с меньшей температурой.

Большинство особей с такими температурами имеют летальный исход.

**В данном случаем, считаю целесообразным:**

Выбросы в большую сторону можно рассматривать в группе с прочими с повышенной температурой.

Выбросы с низкой температурой выделить в отдельную группу, так как мода летальности сильно отлична.

In [284]:
list_outliers(colums_outliers[2])

age  rectal_temper  pulse  respir_rate  temper_limbs  mucous_membr  \
40     9           39.0  150.0         72.0           NaN           NaN   
274    9           38.8  150.0         50.0           1.0           6.0   
54     9           38.6  160.0         20.0           3.0           5.0   
2      9           39.1  164.0         84.0           4.0           6.0   
254    9           38.8  184.0         84.0           1.0           1.0   

     time_fill_capill  pain  abdom_disten  result  
40                NaN   NaN           NaN     1.0  
274               2.0   5.0           2.0     2.0  
54                1.0   3.0           4.0     2.0  
2                 2.0   2.0           4.0     2.0  
254               1.0   4.0           3.0     2.0

#### Комментарий по выбросам данных Пульс (pulse)

Всего 5 выбросов, все с повышенным пульсом и все относятся к молодым особям.

Все, кроме одной (с неполными данными), особи с таким пульсом имеют летальный исход.

Данные коррелируют с данными по возрасту и значимы, но их слишком мало, чтобы на основе их делать выводы.

**В данном случаем, считаю целесообразным:**

Исключить эти данные из выборки

In [285]:
list_outliers(colums_outliers[3])

age  rectal_temper  pulse  respir_rate  temper_limbs  mucous_membr  \
119    1           39.4   54.0         66.0           1.0           2.0   
124    1           38.0   42.0         68.0           4.0           1.0   
264    1            NaN   56.0         68.0           3.0           1.0   
102    9           38.0  140.0         68.0           1.0           1.0   
228    9           38.5  120.0         70.0           NaN           NaN   
294    1            NaN  120.0         70.0           4.0           4.0   
40     9           39.0  150.0         72.0           NaN           NaN   
81     9           38.1  100.0         80.0           3.0           2.0   
207    1           37.8   88.0         80.0           3.0           5.0   
83     1           37.8   60.0         80.0           1.0           2.0   
254    9           38.8  184.0         84.0           1.0           1.0   
2      9           39.1  164.0         84.0           4.0           6.0   
243    9           38.2  124.0         88.0           1.0           2.0   
268    1           39.3   64.0         90.0           2.0           1.0   
185    1           39.3   64.0         90.0           2.0           1.0   
105    1           38.3   52.0         96.0           NaN           3.0   
38     9           39.2  146.0         96.0           NaN           NaN   

     time_fill_capill  pain  abdom_disten  result  
119               1.0   2.0           2.0     1.0  
124               1.0   3.0           2.0     1.0  
264               1.0   3.0           1.0     3.0  
102               1.0   3.0           2.0     1.0  
228               NaN   NaN           NaN     1.0  
294               2.0   2.0           NaN     3.0  
40                NaN   NaN           NaN     1.0  
81                1.0   3.0           1.0     1.0  
207               2.0   NaN           3.0     3.0  
83                2.0   2.0           3.0     1.0  
254               1.0   4.0           3.0     2.0  
2                 2.0   2.0           4.0     2.0  
243               1.0   2.0           4.0     1.0  
268               1.0   NaN           1.0     1.0  
185               1.0   NaN           1.0     1.0  
105               1.0   NaN           NaN     1.0  
38                NaN   NaN           NaN     2.0

#### Комментарий по выбросам данных Частота дыхания (respir_rate)

Всего 17 выбросов, все с повышенной частотой, сильно превосходящей значения 3 квартиля.

Корреляции с другими данными не прослеживается.

Значения выбросов сильно превосходят нормальные значения, и имеют большой разброс, что вызывает сомнения в достоверности этих данных.

**В данном случаем, считаю целесообразным:**

Исключить эти данные из выборки

### Задание 3. Работа с пропусками

1. Рассчитать количество пропусков для всех выбранных столбцов. 

2. Принять и обосновать решение о методе работы с пропусками, 

3. Сформировать датафрейм, в котором пропуски будут отсутствовать.

In [288]:
for column in new_df.columns:
    pct_missing = new_df[column].isnull().mean()
    print(f'{column} - {pct_missing :.1%}')

age - 0.0%
rectal_temper - 20.1%
pulse - 8.0%
respir_rate - 19.4%
temper_limbs - 18.7%
mucous_membr - 15.4%
time_fill_capill - 10.7%
pain - 18.4%
abdom_disten - 18.7%
result - 0.3%


#### Методы работы с пропусками

rectal_temper:
1. Удаление строк - 12 строк из 60 пропусков имеют большое число na, которые стоит удалить, 
2. Игнорирование - не повлияет на анализ данных: пропусков менее 20%
3. Замена на расчетное значение - возможна только после выявления закономерностей

pulse:
1. Удаление строк - 7 строк из 10 пропусков имеют большое число na, которые стоит удалить, 
2. Игнорирование - не повлияет на анализ данных: пропусков менее 8%

respir_rate:
1. Удаление строк - 15 строк из 58 пропусков имеют большое число na, которые стоит удалить,
2. Игнорирование - не повлияет на анализ данных: пропусков менее 20%
3. Замена на расчетное значение - возможна только после выявления закономерностей

temper_limbs:
1. Удаление строк - 27 строк из 56 пропусков имеют большое число na, которые стоит удалить,
2. Игнорирование - не повлияет на анализ данных: пропусков менее 19%
3. Замена на расчетное значение - возможна только после выявления закономерностей

mucous_membr:
1. Удаление строк - 32 строки из 46 пропусков имеют большое число na, которые стоит удалить,
2. Игнорирование - не повлияет на анализ данных: пропусков менее 16%
3. Замена на расчетное значение - возможна только после выявления закономерностей

time_fill_capill
1. Удаление строк - 25 строк из 32 пропусков имеют большое число na, которые стоит удалить,
2. Игнорирование - не повлияет на анализ данных: пропусков менее 11%

pain
1. Удаление строк - 29 строк из 55 пропусков имеют большое число na, которые стоит удалить,
2. Игнорирование - не повлияет на анализ данных: пропусков менее 19%
3. Замена на расчетное значение - возможна только после выявления закономерностей

abdom_disten
1. Удаление строк - 29 строк из 56 пропусков имеют большое число na, которые стоит удалить,
2. Игнорирование - не повлияет на анализ данных: пропусков менее 19%
3. Замена на расчетное значение - возможна только после выявления закономерностей

In [354]:
(new_df[new_df.isnull().T.any().T].
 sort_values('abdom_disten').tail(5))

age  rectal_temper  pulse  respir_rate  temper_limbs  mucous_membr  \
287    1            NaN    NaN          NaN           NaN           NaN   
289    1           38.6   45.0         16.0           2.0           2.0   
292    1            NaN   78.0         24.0           3.0           3.0   
294    1            NaN  120.0         70.0           4.0           4.0   
298    1           37.2   40.0         20.0           NaN           NaN   

     time_fill_capill  pain  abdom_disten  result  
287               NaN   NaN           NaN     1.0  
289               1.0   1.0           NaN     1.0  
292               1.0   NaN           NaN     3.0  
294               2.0   2.0           NaN     3.0  
298               NaN   NaN           NaN     3.0

In [355]:
# удалим все строки, где менее 7 столбцов из 10 заполнены
new_df.dropna(axis = 0, thresh=7).info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 267 entries, 0 to 297
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   age               267 non-null    int64  
 1   rectal_temper     220 non-null    float64
 2   pulse             253 non-null    float64
 3   respir_rate       222 non-null    float64
 4   temper_limbs      241 non-null    float64
 5   mucous_membr      252 non-null    float64
 6   time_fill_capill  260 non-null    float64
 7   pain              240 non-null    float64
 8   abdom_disten      240 non-null    float64
 9   result            266 non-null    float64
dtypes: float64(9), int64(1)
memory usage: 22.9 KB
